net/rpc/server_test.go

```diff
@@ -455,6 +469,56 @@ func TestServeRequest(t *testing.T) {
 	testServeRequest(t, newServer)
 }
 
+func TestServeRequestWithInterceptor(t *testing.T) {
+	beforeHandler := 1
+	afterHandler := 3
+
+	interceptor := ServerServiceCallInterceptor(func(reqServiceMethod string, argv, replyv reflect.Value, handler func()) {
+		// we will assert on this value later
+		beforeHandler = 2
+
+		// these values come from startNewServerWithInterceptor() wiring
+		if reqServiceMethod != "Arith.Add" {
+			t.Errorf("expected serviceMethod in interceptor to be \"Arith.Add\". Was: %s", reqServiceMethod)
+		}
+
+		// argv, replyv reflect.Value,
+		actualArgs := argv.Interface().(Args)
+		if actualArgs.A != 7 || actualArgs.B != 8 {
+			t.Errorf("expected args in interceptor to be {7, 8}. Was: %+v", actualArgs)
+		}
+
+		beforeHandlerReply := replyv.Elem().Interface().(Reply)
+		if beforeHandlerReply.C != 0 {
+			t.Errorf("expected result in interceptor before handler call to be 0. Was %d", beforeHandlerReply.C)
+		}
+
+		// let the RPC req happen
+		handler()
+
+		actualReply := replyv.Elem().Interface().(Reply)
+		if actualReply.C != 15 {
+			t.Errorf("expected result in interceptor to be 15. Was %d", actualReply.C)
+		}
+
+		// we will assert on this value later
+		afterHandler = 4
+
+	})
+
+	startNewServerWithInterceptor(interceptor)
+
+	testServeRequest(t, newServer)
+
+	if beforeHandler != 2 {
+		t.Errorf("expected beforeHandler value to be 2. Was %d", beforeHandler)
+	}
+
+	if afterHandler != 4 {
+		t.Errorf("expected beforeHandler value to be 4. Was %d", afterHandler)
+	}
```

Comment:
I'm not sure how this assert work. Are we sure that the asserts (t.Errorf) inside the interceptor fail properly? if so I'm not sure how we can get into a situation where `beforeHandler=2` and `afterHandler != 4` without the test already failing.

Reply:
Just read @kisunji comment about this and I agree, I think it could be simplified to a bool as checking if handler is called is testing the test.

Reply:
I think we can pursue this in a subsequent PR!